test/utils/tutorials/tester.go

```diff
@@ -0,0 +1,621 @@
+package main
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"os/exec"
+	"path"
+	"strings"
+	"time"
+
+	"github.com/google/shlex"
+	"github.com/google/uuid"
+	. "github.com/onsi/gomega"
+	"github.com/pterm/pterm"
+)
+
+type RunningCommand struct {
+	cmd           *exec.Cmd
+	outb          bytes.Buffer
+	errb          bytes.Buffer
+	cmdPrettyName string
+	cancelFunc    context.CancelFunc
+	stdout        string
+	stderr        string
+}
+
+type ExecutableChunk struct {
+	Stage          string   `json:"stage"`
+	Id             string   `json:"id,omitempty"`
+	Requires       string   `json:"requires,omitempty"`
+	RootDir        string   `json:"rootdir,omitempty"`
+	Runtime        string   `json:"runtime,omitempty"`
+	HereTag        string   `json:"hereTag,omitempty"`
+	Variables      []string `json:"variables,omitempty"`
+	Env            []string `json:"env,omitempty"`
+	Result         []string `json:"result,omitempty"`
+	ReturnCode     int      `json:"return_code,omitempty"` // result of the last command of the chunk
+	IsParallel     bool     `json:"parallel,omitempty"`    //TODO
+	Label          string   `json:"label,omitempty"`
+	HasBreakpoint  bool     `json:"breakpoint,omitempty"`
+	trimedCommands []string
+	rawCommands    []string
+	commands       []*RunningCommand
+}
+
+var dryRun bool = false
+var interactive bool = false
+var verbose bool = false
+var minutesToTimeout int = 10
+var startFrom string = ""
+var operator_root = "./"
+var ingoreBreakpoints bool = false
+var updateTutorials bool = false
+var tutorials_root = "./docs/tutorials"
+
+func initChunk(params string) (*ExecutableChunk, error) {
+	var chunk ExecutableChunk
+	err := json.Unmarshal([]byte(params), &chunk)
+	chunk.trimedCommands = []string{}
+	chunk.ReturnCode = -1
+	if len(chunk.Variables) > 0 {
+		if chunk.Runtime != "bash" {
+			return nil, errors.New("variable extraction are requiring \"runtime\":\"bash\"")
+		}
+	}
+	if chunk.HasBreakpoint {
+		pterm.Warning.Println("breakpoint in the document")
+	}
+	return &chunk, err
+}
+
+func Fail(message string, callerSkip ...int) {
+	panic(message)
+}
+
+func extractStages(tutorial string) [][]*ExecutableChunk {
+	var stages [][]*ExecutableChunk
+	var err error
+	var file *os.File
+	filepath := path.Join(tutorials_root, tutorial)
+	file, err = os.Open(filepath)
+	Expect(err).To(BeNil())
+	defer file.Close()
+	scanner := bufio.NewScanner(file)
+	isInChunk := false
+	searchForEof := false
+	var currentStage string = ""
+	var currentChunk *ExecutableChunk
+	var lineCounter = 0
+	for scanner.Scan() {
+		lineCounter += 1
+		if strings.HasPrefix(scanner.Text(), "```{") {
```

Comment:
JSON is a good language for m2m protocols but for humans could be simpler a command line syntax, i.e.
```
```shell exec --stage="deply" --label=="minikube" --expected-content="created" --expected-result=0
```

The command line parameters could be parsed by using https://pkg.go.dev/flag

Reply:
As we've discussed outside the PR, there are other examples of people using JSON (https://github.com/artemiscloud/activemq-artemis-operator/pull/986).

I'm not against using something else, as long as the text object we get can be parsed and validated against some kind of schema. I don't have another idea than JSON for now.

Feel free to reopen if you get some inspiration :slightly_smiling_face: